File: leet-code/72.编辑距离.js
/*
 * @lc app=leetcode.cn id=72 lang=javascript
 *
 * [72] 编辑距离
 */

// @lc code=start
/**
 * @param {string} word1
 * @param {string} word2
 * @return {number}
 */
var minDistance = function (word1, word2) {
  const dp = Array.from({ length: word1.length + 1 }, () => Array(word2.length + 1).fill(0))
  for (let i = 0; i <= word1.length; i++) {
    dp[i][0] = i
  }
  for (let i = 0; i <= word2.length; i++) {
    dp[0][i] = i
  }
  for (let i = 1; i <= word1.length; i++) {
    const str1 = word1[i - 1];
    for (let j = 1; j <= word2.length; j++) {
      const str2 = word2[j - 1];
      if (str1 === str2) {
        dp[i][j] = dp[i - 1][j - 1]
      } else {
        // 增和删其实都是一样的，
        // 替换要求dp[i-1][j - 1]要相等的前提下 + 1
        dp[i][j] = Math.min(dp[i - 1][j] + 1, dp[i][j - 1] + 1, dp[i - 1][j - 1] + 1)
      }
    }
  }
  return dp[word1.length][word2.length]
};
// @lc code=end

File: leet-code-rs/src/solutions/72_编辑距离.rs
/*
 * @lc app=leetcode.cn id=72 lang=rust
 *
 * [72] 编辑距离
 */

// @lc code=start

#[allow(unused)]
impl Solution {
    /**
     * 解题思路：本题是删与不删问题。
     *  1. dp[i+1][j+1] 定义：s1 前i个元素于 s2 前j个元素的最少操作数
     *  2. 递推公式：
     *     - s1[i] == s2[j]: dp[i+1][j+1] = dp[i][j]
     *     - s1[i] != s2[j]: dp[i+1][j+1] = min(dp[i+1][j],dp[i][j+1],dp[i][j]) + 1
     *       1. s1插入字符：相当于删除 s2，dp[i+1][j+1] = dp[i+1][j] + 1
     *       2. 删除字符：dp[i+1][j+1] = dp[i][j+1] + 1
     *       3. 替换字符：也相当于插入一个和 s2 相等的字符，然后同时删除该字符，dp[i+1][j+1] = dp[i][j] + 1
     *  3. 初始化：dp[i][0] = i， dp[0][j]=j        
     */
    pub fn min_distance(word1: String, word2: String) -> i32 {
        let word1: Vec<char> = word1.chars().collect();
        let word2: Vec<char> = word2.chars().collect();

        let mut dp = vec![vec![500; word2.len() + 1]; word1.len() + 1];
        // 初始化 DP[i][0] 和 DP[0][j]
        for i in 0..=word1.len() {
            dp[i][0] = i;
        }
        for j in 0..=word2.len() {
            dp[0][j] = j;
        }

        for i in 0..word1.len() {
            for j in 0..word2.len() {
                dp[i + 1][j + 1] = if word1[i] == word2[j] {
                    dp[i][j]
                } else {
                    dp[i][j + 1].min(dp[i + 1][j]).min(dp[i][j]) + 1
                }
            }
        }

        return dp[word1.len()][word2.len()] as i32;
    }
}
// @lc code=end

#[allow(unused)]
struct Solution;

#[cfg(test)]
mod test {
    #[allow(unused)]
    use super::*;

    #[test]
    fn test_min_distance() {
        //assert_eq!(Solution::min_distance(vec![]),[]);
        assert!(true)
    }
}
